### ISSUE TITLE
Strategy's hyperopt produce a minor avg profit that the original one

### ISSUE THREAD

**Issue by mb812** [freqtrade/freqtrade#3814]
Hi
My strategy's hyperopt produce a minor avg profit that the original one.
I'm definitely doing something wrong, but I don't find examples for my case.

That's an example:
Let’s take the Strategy002 from freqtrade-startegies.
The buy script look like this:

                (dataframe['rsi'] < 30) &
                (dataframe['slowk'] < 20) &
                (dataframe['bb_lowerband'] > dataframe['close']) &
                (dataframe['CDLHAMMER'] == 100)

Let’s say I want to compare rsi with different values, in a range of 20 - 40.
In the hyperopt file, buy_strategy_generator and indicator_space will look like this:

     @staticmethod
     def buy_strategy_generator(params: Dict[str, Any]) -> Callable:
        def populate_buy_trend(dataframe: DataFrame, metadata: dict) -> DataFrame:
            dataframe.loc[
                (
                  (dataframe['rsi'] < params['rsi-value']) &
                  (dataframe['slowk'] < 20) &
                  (dataframe['bb_lowerband'] > dataframe['close']) &
                  (dataframe['CDLHAMMER'] == 100)
                ),
                'buy'] = 1
            return dataframe
        return populate_buy_trend

    @staticmethod
    def indicator_space() -> List[Dimension]:
        return [
            Integer(20, 40, name='rsi-value'),
            Categorical(['test'], name='trigger')
        ]

Now, if I want to hyperopt my strategy I will need to launch:

    freqtrade hyperopt --strategy Strategy002 --hyperopt Strategy002Hyp --spaces buy --timerange….

The hyperopt suggest me an rsi-value that produce a minor avg profit.
The result should be equal or greater.
What am I doing wrong?

**Comment by xmatthias** [comment]
Well - i would strongly suggest the documentation - it does have some good content where you can learn a ton about this process.

Try for example [this page](https://www.freqtrade.io/en/latest/hyperopt/#loss-functions) - it'll show you that different loss-functions result in different optimization goals (the "default" one optimizes more towards Short holding times, less for profit - so you might have lower profit - but "quicker" trades as a tradeoff).
There is no "best" loss function - as all serve different goals, and opinions on which one is best (or if the best is even included in freqtrade - or still has to be written) will widely vary between users.

I do recommend to read this page from top to bottom when (before?) doing hyperopt.

**Comment by mb812** [comment]
Hi sir, I've read the hyperopt and hyperopt advances guide many times.
The guide suggest that is possible to hyperopt without --strategy command.
But in my case I always get a '--strategy missing' error.
Maybe my problem start here?
I build my hyperopt following the documentation.
But it gives my inefficient rsi-value values.

Should I maybe set 'rsi-value' param in the dataframe in the strategy file too?

**Comment by mb812** [comment]
Correct me if I'm wrong.
Backtesting my strategy in a certain timerange gives me a certain profit. 
The hyperopt of the same startegy in the same timerange should give rsi-value that produce same or better result.

original startegy:
              (dataframe['rsi'] < 30)

hyperopt:
              (dataframe['rsi'] < params['rsi-value']) # 20- 40

Hyperopt should return me rsi-value = 30 if give the best result.
Or another rsi-value if it gives even better results.
Right?

**Comment by xmatthias** [comment]
please point out where you found that it's possible to run hyperopt without strategy. That's not possible since almost a year now - so we should fix this section of the docs.

as said - the use different loss functions (not the default one) should fix it (i did pick the deep-link above intentionally).

**Comment by mb812** [comment]
Ok thank you, now it's clear to me that I can hyperopt only with a strategy.

I cannot really understand why 'better' is relative.
- I'm only testing --spaces buy. This way stoploss and roi table remain unchanged from startegy (stoploss = -0.3, roi = { 0: 0.4 }).
- I'm using the same timerange that I previously used on backtest.
- Hyperopt should simple backtest multiple times the strategy, changing each time the rsi-value from 20 to 40. (look at the hyperopt I wrote above)

So, when the hyperopt will test rsi-value = 30 (after 20, 21, 22, ...., 29), it should produce the same identical trades like in the original backtest.
Right?

**Comment by mb812** [comment]
In the 'execute hyperopt' section of freqtrade.io I found that it's possible to run hyperopt without strategy

     freqtrade hyperopt --config config.json --hyperopt <hyperoptname> -e 500 --spaces all

Probably the strategy is specified in the config file.
Here I had misunderstood.

**Comment by xmatthias** [comment]
because better is relative, also profit wise

* You might be looking for better average returns (at the cost of doing fewer trades) (20 * 2%)
* you might be looking for more trades (at the cost of having a lower average) (40 * 1%)
* you might be looking for lower holding time (which reduces your risk / market exposure).

You'll notice that the first 2 have the same outcome in the end. But which one is better? that's up to you to decide (especially because there's other factors that come into play as well).

Obviously, if you ask me, i'll take 60* 5%  for with the same remaining parameters (timeframe, timerange, pairs...) - but that's not how optimization works - it's always a tradeoff.

---

This is why we have different loss functions - which all optimize in a different direction.

**Comment by xmatthias** [comment]
I'll close this assuming that the above explanation was sufficient to get you going in the right direction.

Feel free to comment below if this is still unclear.
